pystac/extensions/scientific.py

```diff
@@ -0,0 +1,272 @@
+        if not isinstance(other, Doi):
+            return False
+
+        return self.doi.lower() == other.doi.lower()
+
+    def url(self) -> str:
+        return DOI_URL_BASE + parse.quote(self.doi)
+
+
+class Publication:
+    """Helper for Publication entries."""
+    def __init__(self, doi: Union[str, Doi], citation: str) -> None:
+        if not isinstance(doi, Doi):
+            doi = Doi(doi)
+        self.doi = doi
+        self.citation = citation
+
+    def __eq__(self, other: _PublicationType) -> bool:
+        if not isinstance(other, Publication):
+            return NotImplemented
+
+        return self.doi == other.doi and self.citation == other.citation
+
+    def __repr__(self) -> str:
+        return f'<Publication doi={self.doi} target={self.citation}>'
+
+    def to_dict(self) -> Dict[str, str]:
+        return copy.deepcopy({'doi': self.doi.doi, 'citation': self.citation})
+
+    @staticmethod
+    def from_dict(d: Dict[str, str]) -> _PublicationType:
+        return Publication(d['doi'], d['citation'])
+
+    def get_link(self) -> link.Link:
+        url = self.doi.url()
+        return link.Link(CITE_AS, url)
+
+
+def remove_link(links: List[link.Link], pub: Publication):
+    url = pub.doi.url()
+    for i, a_link in enumerate(links):
+        if a_link.rel != CITE_AS:
+            continue
+        if a_link.target == url:
+            del links[i]
+            break
+
+
+class ScientificItemExt(base.ItemExtension):
+    """Add an citation and dois to a STAC Item."""
+    def __init__(self, an_item: item.Item) -> None:
+        self.item = an_item
+
+    def apply(self,
+              doi: Optional[Union[str, Doi]] = None,
+              citation: Optional[str] = None,
+              publications: Optional[List[Publication]] = None) -> None:
+        if doi:
+            self.doi = doi
+        if citation:
+            self.citation = citation
+        if publications:
+            self.publications = publications
+
+    @classmethod
+    def from_item(cls: _ScientificItemExtType, an_item: item.Item) -> _ScientificItemExtType:
+        return cls(an_item)
+
+    @classmethod
+    def _object_links(cls: _ScientificItemExtType) -> List:
+        return []
+
+    @property
+    def doi(self) -> Doi:
+        return self.item.properties.get(DOI)
+
+    @doi.setter
+    def doi(self, v: Union[str, Doi]) -> None:
+        if isinstance(v, Doi):
+            v = v.doi
+        self.item.properties[DOI] = v
+        url = DOI_URL_BASE + self.doi
+        self.item.add_link(pystac.link.Link(CITE_AS, url))
+
+    @property
+    def citation(self) -> str:
+        return self.item.properties.get(CITATION)
+
+    @citation.setter
+    def citation(self, v: str) -> None:
+        self.item.properties[CITATION] = v
+
+    @property
+    def publications(self) -> List[Publication]:
+        return [Publication.from_dict(pub) for pub in self.item.properties.get(PUBLICATIONS, [])]
```

Comment:
Didn't catch this in my last review, but:

The way that Publications are handled here breaks a mutability pattern in PySTAC.

For instance, when I do

```python
item.ext.scientific.doi = "something"
print(item.ext.scientific.doi)
```

I would get the expected result, "something". This is true even though the extension wrapper object is created twice - once for each call of `item.ext.scientific`.

As it is currently written, the same wouldn't hold true for publications. For example, say I had an item with a single publication that had a citation of "everyone et al". The following:

```python
item.ext.scientific.publications[0].citation = "none et al"
print(item.ext.scientific.publications[0].citation
```

would print "everyone et al" - since the Publication instance that was returned was a temporary object, setting it's citation didn't have consequence.

This could be worked around by requiring the user always get and set publications, e.g.

```python
pubs = item.ext.scientific.publications
pubs[0].citation = "none et al"
item.ext.scientific.publications = pubs
```

but this is burdensome, and surely users will forget to do this. Also, it doesn't match the expectation of mutability that's present elsewhere in the library, including the Ext classes that wrap an item rather than contain data of their own.

The way around this is to keep the wrapper train rolling - take [LabelOverview](https://github.com/stac-utils/pystac/blob/e105c4cb0bdd5f1273b85fda9828edce80b714c4/pystac/extensions/label.py#L380) for instance. It takes as a constructor the dict from the parent Item's properties and uses getters and setters to manipulate the data directly on the Item rather than hold its own state.

I think we should follow that pattern here and provide similar wrappers around Item state with the Publications  and DOI objects.

Reply:
I'll take a look at this today.  Now that you say this, I realize that version is likely broken.  Created #210.

Reply:
I'm having trouble thinking through how best to do this.  I'd like Publication and Doi instances to be able to live outside of an Item or Collection.  What should a Publication track?  I see a couple possibilities with it being optional if these are attached:

* The Item or Collection
* The properties/extra_fields and links
* The specific entries in properties/extra_fields and the link

Having a reference to the specific dict in `sci:publications` and in links could make changing the doi or citation fairly easy to do, but removing publications from an Item/Collection becomes trouble, yes?

I do worry about things like people taking a publication for one item and then adding that publication to all the other items in a collection and trying to permute them to match that item.  If there are references inside the publication, then all the copies will end up changing together.  And now I'm thinking which things should have deepcopies and which shouldn't getting all turned around.

I don't have an understand of what the overall strategy is for mutations.

Reply:
I'd suggest wrapping the dictionary of properties that make up the Publication, i.e. the __init__ takes a JSON serialized version of the Publication. In other cases, I have a `create` classmethod which acts as an overloaded constructor that takes the same params as `apply` and allows the objects to live outside of an Item (e.g. [how it's done with Bands](https://github.com/stac-utils/pystac/blob/1fcfb44220965f230e73f89ee9a974b346e67351/tests/extensions/test_eo.py#L37))

I'm not sure that removing publications is too onerous, the logic in the removal would have to just account for the fact it's removing a dict from a list rather than an object from the list.

The scenario you mention is problematic, and a good flag to raise. If you were to add the publication from one item to another and change the publication instance, that change would appear in both places. This could lead to unexpected behavior. The way I can see coding around this case is to have these objects track their owners, much like a link or asset tracks their owners. That way, when a publication owned by one item is attached to another, either the ownership can switch or a deepcopy made.

It does seem to me that this scenario is less likely to cause confusion than if the publications become detached and are un-editable unlike the rest of the library. I think if we're clear that things are highly mutable, users should be careful about reusing instances set on many different objects and then editing those instances, and use a copy themselves if they want a fresh copy that won't be mutated by the original instance. It seems similar to the scenario of  a user creating a single `dict` and assigning it to many Item instance's `properties` via the constructor, and then mutating the properties. So, the scenario you outline is problematic, but I think is something we can live with. 

I'm not sure what the overall strategy for mutations is, but I'll try to comment on my perspective around mutability in PySTAC. Mutability always comes with risk of state modification when you don't want it, and ideally all things would be strictly immutable. Any change would give you back a new copy of the thing you changed, with the change you requested. However, this gets complex pretty quickly in STAC. Imaging you've read in a large catalog. Any change in that catalog would produce a new catalog with the relevant edits. If I modify a single property of an Item, then I need a new Item. I then need all of the links to that Item changed, and so on - what I really would produce is an entirely new STAC. Things get more complicated when you consider that PySTAC uses caching to enable links to the same STAC object on the filesystem to refer to the same STAC object in memory. 

Immutability is possible, and there are patterns like Lenses to make complex hierarchical immutable data structures easier to transform. When I was setting things up in PySTAC, I certainly considered this. However, Python being the dynamic, mutation-friendly language that it is, that felt like it would be swimming against the current of what users would want. I think of STACs as graphs, so I looked at the most popular graph library - networkx - to take direction from, which has mutable data structures (as an aside, I found this [interesting conversation about mutable graphs](https://github.com/dmlc/dgl/issues/33) that is related).

So to me, if we're not trying to be immutable, then we can lean into mutability and use it to our advantage, like how we allow multiple extension objects act as views into the STAC object properties and mutate it, so that the extension objects are essentially stateless. We can also ask the user to assume mutability and understand that there's risk in assigning the same instance across multiple STAC objects.

Reply:
Looking at the code, I realized that `Doi` is not an object, but a wrapper around a string. This would be fine to keep as an object, though I think it may get awkward to do things like

```
item.ext.scientific.doi.doi
```
to get at the string. Perhaps consider moving the logic into the scientific extensions? E.g. I could imagine validation logic and the URL being directly on ScientificItemExt, like a `doi_url` property that is only a getter, and a `validate_doi` method that is called on the doi setter, which would take a string.

For `Publication`, the above comments hold - this is a object that would be represented as a dict inside the item properties.

Reply:
Should the publication class go away and just use a dict?  I thought python3 had a [dotted dict](https://stackoverflow.com/questions/2352181/how-to-use-a-dot-to-access-members-of-dictionary) type thing, but ?  If we wanted to ditch mutability of the fields inside the pub, we could use [collections.namedtuple](https://docs.python.org/3/library/collections.html#collections.namedtuple)

Reply:
After taking another look at this and pushing some code around, I think this is a tricker case than other cases, for instance `LabelClasses`, because this is a list of objects and not an dict. The 'proper' way to do what I'm talking about is to generate a collection class that would define it's own sequence methods, that would translate between the Item properties and Publication objects as items were retrieved, appended etc.

However, this adds a level of complexity such that the "worth it" factor is starting to tip in the direction of "no" for me. The way you have implemented works great, and the situation I mention may come up and confuse users but perhaps won't be too much of head-scratcher to work around.

That's to say, let's go forward with this solution and maybe re-evaluate if someone complains. Thanks for bearing with me on this!